File: apps/web/src/lib/story-display.ts
/*
 * Visible story formatting. Same strip set as plain-text.ts (ATX / bullets / stars),
 * but newlines stay — the detail panel renders paragraphs. Never strip `_`.
 */

const ATX = /^\s{0,3}#{1,6}\s+/gm;
const BULLET = /^\s{0,3}([-*+]|\d{1,9}[.)])\s+/gm;

export function stripVisibleMarkdown(text: string): string {
  return text
    .replace(ATX, "")
    .replace(BULLET, "")
    .replace(/\*\*/g, "")
    .replace(/\*/g, "");
}

export function formatStoryForDisplay(text: string | null | undefined): string | null {
  if (typeof text !== "string") return null;
  const cleaned = stripVisibleMarkdown(text).replace(/[ \t]+\n/g, "\n").trim();
  return cleaned.length > 0 ? cleaned : null;
}

export function storyParagraphs(text: string | null | undefined): string[] {
  const formatted = formatStoryForDisplay(text);
  if (!formatted) return [];
  return formatted.split(/\n\s*\n/).map((part) => part.trim()).filter(Boolean);
}

File: pipelines/story_display.py
"""Visible story strip — same rules as apps/web/src/lib/story-display.ts.

Bake / live-db / gate must call this, not a second copy. Never strip `_`.
"""

from __future__ import annotations

import re

_ATX = re.compile(r"^\s{0,3}#{1,6}\s+", re.M)
_BULLET = re.compile(r"^\s{0,3}([-*+]|\d{1,9}[.)])\s+", re.M)


def strip_visible_markdown(text: str) -> str:
    cleaned = _ATX.sub("", text)
    cleaned = _BULLET.sub("", cleaned)
    return cleaned.replace("**", "").replace("*", "")


def format_story_for_display(text: str | None) -> str | None:
    if not isinstance(text, str):
        return None
    cleaned = re.sub(r"[ \t]+\n", "\n", strip_visible_markdown(text)).strip()
    return cleaned or None
